extensions/data-plane/data-plane-google-bigquery/src/main/java/org/eclipse/edc/connector/dataplane/gcp/bigquery/pipeline/BigQueryDataSink.java

```diff
@@ -0,0 +1,337 @@
+                    var element = mapper.readTree(jsonParser);
+                    var jsonString = element.toString();
+                    var page = new JSONArray(jsonString);
+                    append(page);
+                    token = jsonParser.nextToken();
+                }
+
+                if (token != null && token != JsonToken.START_ARRAY) {
+                    throw new IllegalStateException("BigQuery sink JSON array expected");
+                }
+            } catch (Exception exception) {
+                errorWhileAppending = true;
+                appendException = exception;
+                monitor.severe("BigQuery Sink error while appending: ", appendException);
+                break;
+            }
+
+            if (part instanceof BigQueryPart bigQueryPart) {
+                if (bigQueryPart.getException() != null) {
+                    errorWhileAppending = true;
+                    appendException = bigQueryPart.getException();
+                    monitor.severe("BigQuery Sink error from source: ", appendException);
+                    break;
+                }
+            }
+        }
+
+        inflightRequestCount.arriveAndAwaitAdvance();
+        closeSinkStream();
+
+        monitor.debug("BigQuery Sink requests arrived, stream " + streamWriter.getStreamName() + " for table " + target.getTableName().toString() + " closed, now finalizing..");
+        var finalizeResponse =
+                writeClient.finalizeWriteStream(streamWriter.getStreamName());
+        monitor.debug("BigQuery Sink rows written: " + finalizeResponse.getRowCount());
+
+        // Commit the data stream received only if the stream writer succeeded (transaction).
+        // TODO support option to enable commit even if the stream writer failed.
+        if (!errorWhileAppending) {
+            commitData();
+        }
+
+        writeClient.shutdownNow();
+        var waitIterationCount = 0;
+        var terminated = false;
+        do {
+            try {
+                terminated = writeClient.awaitTermination(2, TimeUnit.SECONDS);
+            } catch (InterruptedException interruptedException) {
+                monitor.warning("BigQuery Sink write client shut down (interrupted)");
+            }
+            waitIterationCount++;
+        } while (!terminated && waitIterationCount < 3);
+
+        if (terminated) {
+            monitor.info("BigQuery Sink write client shut down");
+        } else {
+            monitor.warning("BigQuery Sink write client NOT shut down after timeout");
+        }
+
+        if (errorWhileAppending) {
+            if (appendException != null) {
+                return failure(new StreamFailure(List.of("BigQuery Sink error :" + appendException), GENERAL_ERROR));
+            }
+            return failure(new StreamFailure(List.of("BigQuery Sink error"), GENERAL_ERROR));
+        }
+
+        return success();
+    }
+
+    private void checkStreamWriter()
```

Comment:
pls put exception declarations on the same line.

More importantly, I think this method is spurious in that it does too many things. IMO, there should be a distinct point where the stream is opened (i.e. in `transferParts()`) and a distinct point where it's closed (e.g. in a `finally` statement). 
This would also make the `streamWriter` a local variable in the `transferParts` method, and not a field.

In testing, the `streamWriter` would have to be mocked indirectly via `writeClient -> writeStream`

Reply:
Done